### PR TITLE
Allow macs to run --rely_on_ssh_auth with docker

### PR DESCRIPTION
**Pull request by nik9000**
Macs can't forward the ssh authentication socket used by `--rely_on_ssh_auth`. This implements a mac-only, alternate, and slightly more janky method of sharing the authentication: it runs openssh-server in the docker container and `ssh`es with `-A` to forward the socket.

I justify it this way:
1. Folks with macs *do* have to be able to run these builds.
2. Removing this will allow me to drop support for http based authentication (#584). I've not used another project that requires it and it is difficult to set up compared to the ssh authentication that we all set up already.
3. CI won't use this. That is bad because it won't get as much testing, but it is good because the ssh-based method can be written as though it'll only be run in the foreground.
4. The docker for mac folks have been [working](https://github.com/docker/for-mac/issues/410) on "something" to fix this, but they haven't got there in years. So maybe this problem will just vanish one day?

### CODE REVIEW

**Comment by nik9000** [comment]
This solution is similar in spirit to https://github.com/avsm/docker-ssh-agent-forward but it doesn't require a separate docker container that you have to manage so I don't have to worry about leaving that container running accidentally. That solution is good, but for the "docker is transparent" nature of this project I like this one better. They are all fairly bad and docker for mac should build something better into docker itself but since I'm not about to go and do it myself I really can't complain too much.

**Comment by nik9000** [comment]
I have tested this a bunch on my mac. I haven't tested it with python 3 and I haven't put it through pep8 or tested it on my linux machine. So it isn't ready to be merged. But it is ready to be looked at.

**Comment by nik9000** [comment]
It also doesn't allow my mac to succeed with `./build_docs --all --rely_on_ssh_auth --target_repo git@github.com:elastic/built-docs-nik.git --reference ~/.git-references/`. I believe I'll have to build an option to move the `.repos` directory out of the source directory. But that is kind of a parallel thing.

**Comment by nik9000** [comment]
> It also doesn't allow my mac to succeed with

I think this is worth explaining: the process fails with `Too many open files in system` when it tries to check out the built-docs repo. I suspect this is because it checks it out into the same share as the code which is marked `cached`. If it checked it out into an actual tmp directory I think we're be fine. But one thing at a time!

**Comment by nik9000** [comment]
> I have tested this a bunch on my mac. I haven't tested it with python 3 and I haven't put it through pep8 or tested it on my linux machine. So it isn't ready to be merged. But it is ready to be looked at.

I've got it working in whatever version of python 3 is on my mac and passing the pep8 it has. I'll likely do linux testing in the morning.

**Comment by nik9000** [comment]
> I've got it working in whatever version of python 3 is on my mac and passing the pep8 it has. I'll likely do linux testing in the morning.

Linux looks fine for me too.

@jarpy or @ddillinger, could you have a look at this one?

File: build_docs.pl [hunk 1]
```diff
@@ -1,5 +1,8 @@
 #!/usr/bin/env perl
 
+# Flush on every print even if we're writing to a pipe (like docker).
+$| = 1;
+
```
Comment (jarpy) [inline build_docs.pl]: Nice. Pity we can't set line-buffering so consistently in Python.

File: build_docs.pl [hunk 2]
```diff
@@ -616,10 +635,15 @@ sub init_env {
                 use POSIX ":sys_wait_h";
                 my $child_status = 'missing';
                 while ((my $child = waitpid(-1, WNOHANG)) > 0) {
-                    $child_status = $? if ( $child == $child_pid );
+                    my $status = $? >> 8;
```
[the remark this answers is not in the document]
Comment (jarpy) [reply build_docs.pl]: No worries. Is that bit-shifting!?

File: Dockerfile [hunk 3]
```diff
@@ -17,16 +34,17 @@ RUN install_packages \
...
   python \
   ruby \
   ruby-dev \
   unzip \
   xsltproc
 
-# We mount this log directory as tmp directory so we can't have
-# files there.
-RUN rm -rf /var/log/nginx
+# We mount these directories with tmpfs so we can write to them while so they
```
Comment (jarpy) [inline Dockerfile]: Slight jumble of words here.

File: build_docs.pl (hunk 2 above)
[the remark this answers is not in the document]
Comment (nik9000) [reply build_docs.pl]: It is! It is how perl digs up the actual exit code of the child. I expect it is passing on the c standard behavior though.

**Comment by nik9000** [comment]
Thanks for reviewing @jarpy!

**Pull request merged**